### DISCUSSION TITLE
Token with "repo" scope to install private packages?

### DISCUSSION THREAD

**Discussion by lomedil** [community/community#27070] [Actions]
<p>I feel some confused about tokens and scopes when used to share packages from private repositories.</p>
<p>I have a private repository with some collaborators. I want “publish” a package from this repo to be shared among those collaborators. According to “<a href="https://help.github.com/en/packages/publishing-and-managing-packages/about-github-packages#about-tokens" rel="nofollow noopener">About Github Packages page</a>”, I need to create a token with “read:packages” and “repo” scopes because my repo is private. So, with that token my collegues could log in Github npm registry and install the package.</p>
<p>But the token has full access to all my private repositories because “repo” scope is flagged. Is that right? Is there any other way to share packages from private repositories?</p>
<p>Thanks in advance for your time.</p>

**Comment by jcansdale** [comment, marked as answer]
<p>Hi <a class="mention" href="/lomedil">@lomedil</a>,</p>
<blockquote>
<p>According to “<a href="https://help.github.com/en/packages/publishing-and-managing-packages/about-github-packages#about-tokens" rel="nofollow noopener">About Github Packages page </a>”, I need to create a token with “read:packages” and “repo” scopes because my repo is private.</p>
</blockquote>
<p>I don’t think our documentation is correct here. You’ll find you only need the <code>read:packages</code> scope to install private packages.</p>
<blockquote>
<p>But the token has full access to all my private repositories because “repo” scope is flagged. Is that right? Is there any other way to share packages from private repositories?</p>
</blockquote>
<p>What I suggest you do is create a <a href="https://developer.github.com/v3/guides/managing-deploy-keys/#machine-users" rel="nofollow noopener">machine-user</a> account that has access to the private repositories you need to share packages from (it only needs <code>read</code> access). You can then generate PATs with the <code>read:package</code> scope from this account.</p>
<p>If this PAT might be intentionally pushed to a public repository, you will need to encode it to prevent it from being automatically deleted by GitHub.</p>
<p>I’ve created a tool that will let you encode <code>read:packages</code> PATs for use in various package ecosystems. If you have Docker installed, you can use it like this:</p>
<pre><code class="lang-auto">docker run jcansdale/gpr encode &lt;READ_PACKAGES_TOKEN&gt;
</code></pre>
<p>It will output the following:</p>
<pre><code class="lang-auto">An encoded token can be included in a public repository without being automatically deleted by GitHub.
These can be used in various package ecosystems like this:

A NuGet `nuget.config` file:
&lt;packageSourceCredentials&gt;
  &lt;github&gt;
    &lt;add key="Username" value="PublicToken" /&gt;
    &lt;add key="ClearTextPassword" value="&amp;#60;&amp;#82;&amp;#69;&amp;#65;&amp;#68;&amp;#95;&amp;#80;&amp;#65;&amp;#67;&amp;#75;&amp;#65;&amp;#71;&amp;#69;&amp;#83;&amp;#95;&amp;#84;&amp;#79;&amp;#75;&amp;#69;&amp;#78;&amp;#62;" /&gt;
  &lt;/github&gt;
&lt;/packageSourceCredentials&gt;

A Maven `settings.xml` file:
&lt;servers&gt;
  &lt;server&gt;
    &lt;id&gt;github&lt;/id&gt;
    &lt;username&gt;PublicToken&lt;/username&gt;
    &lt;password&gt;&amp;#60;&amp;#82;&amp;#69;&amp;#65;&amp;#68;&amp;#95;&amp;#80;&amp;#65;&amp;#67;&amp;#75;&amp;#65;&amp;#71;&amp;#69;&amp;#83;&amp;#95;&amp;#84;&amp;#79;&amp;#75;&amp;#69;&amp;#78;&amp;#62;&lt;/password&gt;
  &lt;/server&gt;
&lt;/servers&gt;

An npm `.npmrc` file:
@OWNER:registry=https://npm.pkg.github.com
//npm.pkg.github.com/:_authToken="\u003c\u0052\u0045\u0041\u0044\u005f\u0050\u0041\u0043\u004b\u0041\u0047\u0045\u0053\u005f\u0054\u004f\u004b\u0045\u004e\u003e"
</code></pre>
<p>Your collaborators can this this to access your private packages from their public or private repositories.</p>
<p>Because there is the possibility that someone might decode one of these tokens and push it to a public repository, to makes sense to create a new PAT for each set of collaborators.</p>
<p>Does that help at all?</p>

**Comment by lomedil** [comment]
<p>Great! I have created a <em>machine-user</em> account, added as collaborator for one of my private proyects and created a PAT. Now with that PAT I can donwload packages from that private repository. This <em>strategy</em> solves this and another questions that I was thinking about.</p>
<p>I didn’t try your Docker application because I don’t have installed it on this machine. But I know why a PAT must be encoded and how I can do it.</p>
<p>About ‘repo’ scope for PAT. I afraid that docs are right. When I try to login to NPM using a PAT with just the ‘read:packages’ scope, it throws this error:</p>
<pre><code class="lang-shell">npm ERR! code E401
npm ERR! 401 Unauthorized - PUT https://npm.pkg.github.com/-/user/org.couchdb.user:lomedil - Error authenticating user: Personal Access Token is invalid. Your token must have the `repo` and read:packages` scopes to login to the GitHub Package Registry.
</code></pre>
<p>Anyway. Your answer are more than correct. Thank you very much.</p>

**Comment by jcansdale** [comment]
<p><a class="mention" href="/lomedil">@lomedil</a>,</p>
<blockquote>
<p>About ‘repo’ scope for PAT. I afraid that docs are right. When I try to login to NPM using a PAT with just the ‘read:packages’ scope, it throws this error…</p>
</blockquote>
<p>Instead of using <code>npm login</code>, could you try creating a <code>.npmrc</code> at the root of your repository (changing <code>OWNER</code> to your user/org):</p>
<pre><code class="lang-auto">@OWNER:registry=https://npm.pkg.github.com
//npm.pkg.github.com/:_authToken=${GITHUB_TOKEN}
</code></pre>
<p>Then do:</p>
<pre><code class="lang-auto">export GITHUB_TOKEN=&lt;your read:packages PAT&gt;
npm install
</code></pre>
<p>Could you let me know if this works?</p>

**Comment by lomedil** [comment]
<p>Yes, you are right. To be more specific, I tried to install a global package, not a module as project dependency. Based in your suggestion, I created a file <code>~/.npmrc</code> using a PAT with <em>read:package</em> and now it works.</p>
<p>Once more, thank you very much.</p>

**Comment by danurahadi** [comment]
<p>Hello <a class="mention" href="/jcansdale">@jcansdale</a> ,</p>
<p>I’ve tried your tool to encode my PAT and this is the result i got :</p>
<pre><code class="lang-auto">docker run jcansdale/gpr encode my_token_abcd                  

An encoded token can be included in a public repository without being automatically deleted by GitHub.
These can be used in various package ecosystems like this:

A NuGet `nuget.config` file:
&lt;packageSourceCredentials&gt;
  &lt;github&gt;
    &lt;add key="Username" value="PublicToken" /&gt;
    &lt;add key="ClearTextPassword" value="&amp;#109;y_token_abcd" /&gt;
  &lt;/github&gt;
&lt;/packageSourceCredentials&gt;

A Maven `pom.xml` file:
&lt;repositories&gt;
  &lt;repository&gt;
    &lt;id&gt;github-public&lt;/id&gt;
    &lt;url&gt;https://public:&amp;#109;y_token_abcd@maven.pkg.github.com/&lt;OWNER&gt;/*&lt;/url&gt;
  &lt;/repository&gt;
&lt;/repositories&gt;

An npm `.npmrc` file:
@OWNER:registry=https://npm.pkg.github.com
//npm.pkg.github.com/:_authToken="\u006dy_token_abcd"
</code></pre>
<p>I use your docker image latest version, Docker version 20.10.7, and running on Ubuntu 20.04.<br>
It seems like something went wrong in the encoded token.</p>
<p>Is it a bug on your tool or am I missing something here?</p>
<p>Thank you, Argi.</p>
